beacon-chain/state/stategen/replayer.go

```diff
@@ -0,0 +1,347 @@
+		}
+		root, err := b.Block().HashTreeRoot()
+		if err != nil {
+			// use this error message to wrap a sentinel error for error type matching
+			wrapped := errors.Wrap(ErrInvalidDBBlock, err.Error())
+			msg := fmt.Sprintf("could not compute hash_tree_root for block @ slot=%d", b.Block().Slot())
+			return [32]byte{}, nil, errors.Wrap(wrapped, msg)
+		}
+		canon, err := r.c.IsCanonical(ctx, root)
+		if err != nil {
+			return [32]byte{}, nil, errors.Wrap(err, "error from blockchain info fetcher IsCanonical check")
+		}
+		if canon {
+			return root, b, nil
+		}
+	}
+	return [32]byte{}, nil, errors.Wrap(ErrNoCanonicalBlockForSlot, "no good block for slot")
+}
+
+// ancestorChain works backwards through the chain lineage, accumulating blocks and checking for a saved state
+// if it finds a saved state the the tail block was descended from, it returns this state and
+// all blocks in the lineage, including the tail block. blocks are returned in ascending order.
+// note that this function assumes that the tail is a canonical block, and therefore assumes that
+// all ancestors are also canonical.
+func (r *canonicalChainer) ancestorChain(ctx context.Context, tail block.SignedBeaconBlock) (state.BeaconState, []block.SignedBeaconBlock, error) {
+	ctx, span := trace.StartSpan(ctx, "canonicalChainer.ancestorChain")
+	defer span.End()
+	chain := make([]block.SignedBeaconBlock, 0)
+	for {
+		if err := ctx.Err(); err != nil {
+			msg := fmt.Sprintf("context canceled while finding ancestors of block @ slot %d", tail.Block().Slot())
+			return nil, nil, errors.Wrap(err, msg)
+		}
+		b := tail.Block()
+		// compute hash_tree_root of current block and try to look up the corresponding state
+		root, err := b.HashTreeRoot()
+		if err != nil {
+			msg := fmt.Sprintf("could not compute htr for descendant block @ slot=%d", b.Slot())
+			return nil, nil, errors.Wrap(err, msg)
+		}
+		st, err := r.h.StateOrError(ctx, root)
+		// err from StateOrError == nil, we've got a real state - the job is done!
+		if err == nil {
+			// this handles cases where we find the state right away. return empty list of blocks
+			// we found the state by the root of the head, meaning it has already been applied.
+			// we only want to return the blocks descended from it.
+			reverseChain(chain)
+			return st, chain, nil
+		}
+		// ErrNotFoundState errors are fine, but other errors mean something is wrong with the db
+		if !errors.Is(err, db.ErrNotFoundState) {
+			return nil, nil, errors.Wrap(err, fmt.Sprintf("error querying database for state w/ block root = %#x", root))
+		}
+		parent, err := r.h.Block(ctx, bytesutil.ToBytes32(b.ParentRoot()))
+		if err != nil {
+			msg := fmt.Sprintf("db error when retrieving parent of block @ slot=%d by root=%#x", b.Slot(), b.ParentRoot())
+			return nil, nil, errors.Wrap(err, msg)
+		}
+		if helpers.BeaconBlockIsNil(parent) != nil {
+			msg := fmt.Sprintf("unable to retrieve parent of block @ slot=%d by root=%#x", b.Slot(), b.ParentRoot())
```

Comment:
```suggestion
			msg := fmt.Sprintf("unable to retrieve parent of block at slot=%d by root=%#x", b.Slot(), b.ParentRoot())
```

Reply:
I don't feel strongly about this (saving 1 whole character :partying_face: ) but is there some specific concern w/ using `@` as shorthand for `at`?

Reply:
My concern is that it's less readable, especially when you have other special chars in the message.